src/main/controllers/ledger-service.js

```diff
@@ -1,151 +1,100 @@
 'use strict';
 
-const electron = require('electron');
-const ledgerco = require('ledgerco');
-const EthereumTx = require('ethereumjs-tx');
-const { timeout, TimeoutError } = require('promise-timeout');
-
-const CONFIG = require('../config');
+const electron = require('electron'),
+	{ timeout, TimeoutError } = require('promise-timeout'),
+	Web3 = require('web3'),
+	ProviderEngine = require('web3-provider-engine'),
+	FetchSubprovider = require('web3-provider-engine/subproviders/fetch'),
+	Transport = require('@ledgerhq/hw-transport-node-hid').default,
+	createLedgerSubprovider = require('@ledgerhq/web3-subprovider').default,
+	NETWORK_ID = require('../config').chainId,
+	SELECTED_SERVER_URL = require('./web3-service').SELECTED_SERVER_URL;
+
+const errorsMap = {
+	'ledger device: unknown_error (0x6801)': 'LEDGER_IS_TIMED_OUT'
+};
 
 module.exports = function(app) {
-	let comm = null;
-	let lastHIDPath = '';
-
-	const getNetworkId = () => {
-		return CONFIG.chainId;
-	};
-
-	const getDefaultDerivationPath = () => {
-		return "44'/60'/0'/0";
-	};
-
 	const controller = function() {};
 
-	async function closeConnection() {
-		if (!comm) {
-			return;
-		}
+	function createWeb3(accountsOffset, accountsQuantity) {
+		accountsOffset = accountsOffset || 0;
+		accountsQuantity = accountsQuantity || 6;
+		const engine = new ProviderEngine();
+		const getTransport = () => Transport.create();
+		const ledger = createLedgerSubprovider(getTransport, {
+			networkId: NETWORK_ID,
+			accountsLength: accountsQuantity,
+			accountsOffset
+		});
 
-		await comm.close_async();
-		comm = null;
+		engine.addProvider(ledger);
+		engine.addProvider(new FetchSubprovider({ rpcUrl: SELECTED_SERVER_URL }));
+		engine.start();
+		return new Web3(engine);
 	}
-	async function getEth() {
-		let list = await ledgerco.comm_node.list_async();
-		let isConnected = list && list.length;
-
-		if (isConnected) {
-			let newHDPath = list[0];
-			if (comm && lastHIDPath && lastHIDPath != newHDPath) {
-				await closeConnection();
-			}
-			lastHIDPath = newHDPath;
-		} else {
-			throw Error('Device not found');
-		}
 
-		if (!comm) {
-			comm = await ledgerco.comm_node.create_async();
+	let onError = (err, reject) => {
+		if (!reject) {
+			return;
 		}
 
-		return new ledgerco.eth(comm);
-	}
-
-	function obtainPathComponentsFromDerivationPath(derivationPath) {
-		// check if derivation path follows 44'/60'/x'/n pattern
-		const regExp = /^(44'\/6[0|1]'\/\d+'?\/)(\d+)$/;
-		const matchResult = regExp.exec(derivationPath);
-		if (matchResult === null) {
-			throw new Error(
-				"To get multiple accounts your derivation path must follow pattern 44'/60|61'/x'/n "
-			);
+		if (err instanceof TimeoutError) {
+			return reject('CUSTOM_TIMEOUT');
```

Comment:
If `reject` is a promise rejection we should only ever pass `Error` objects into it.